cirq/google/calibration/engine_simulator.py

```diff
@@ -0,0 +1,405 @@
+        for request in requests:
+            if isinstance(request, FloquetPhasedFSimCalibrationRequest):
+                options = request.options
+                characterize_theta = options.characterize_theta
+                characterize_zeta = options.characterize_zeta
+                characterize_chi = options.characterize_chi
+                characterize_gamma = options.characterize_gamma
+                characterize_phi = options.characterize_phi
+            else:
+                raise ValueError(f'Unsupported calibration request {request}')
+
+            translated_gate = self._gates_translator(request.gate)
+            if translated_gate is None:
+                raise ValueError(f'Calibration request contains unsupported gate {request.gate}')
+
+            parameters = {}
+            for a, b in request.pairs:
+                drifted = self._create_gate(a, b, translated_gate)
+                parameters[a, b] = PhasedFSimCharacterization(
+                    theta=drifted.theta if characterize_theta else None,
+                    zeta=drifted.zeta if characterize_zeta else None,
+                    chi=drifted.chi if characterize_chi else None,
+                    gamma=drifted.gamma if characterize_gamma else None,
+                    phi=drifted.phi if characterize_phi else None,
+                )
+
+            results.append(
+                PhasedFSimCalibrationResult(
+                    parameters=parameters, gate=request.gate, options=options
+                )
+            )
+
+        return results
+
+    def _run(
+        self, circuit: Circuit, param_resolver: ParamResolver, repetitions: int
+    ) -> Dict[str, np.ndarray]:
+        converted = self._convert_to_circuit_with_drift(circuit)
+        return self._simulator._run(converted, param_resolver, repetitions)
+
+    def _base_iterator(
+        self,
+        circuit: Circuit,
+        qubit_order: QubitOrderOrList,
+        initial_state: Any,
+    ) -> Iterator[StepResult]:
+        converted = self._convert_to_circuit_with_drift(circuit)
+        return self._simulator._base_iterator(converted, qubit_order, initial_state)
+
+    def _convert_to_circuit_with_drift(self, circuit: Circuit) -> Circuit:
+        copied = Circuit(circuit)
+        converter = self._PhasedFSimConverter(self)
+        converter.optimize_circuit(copied)
+        return copied
+
+    def _create_gate(self, a: Qid, b: Qid, gate: FSimGate) -> PhasedFSimGate:
+        if (a, b, gate) in self._drifted_parameters:
+            parameters = self._drifted_parameters[(a, b, gate)]
+        elif (b, a, gate) in self._drifted_parameters:
+            parameters = self._drifted_parameters[(b, a, gate)].parameters_for_qubits_swapped()
+        else:
+            parameters = self._drift_generator(a, b, gate)
+            self._drifted_parameters[(a, b, gate)] = parameters
+        return PhasedFSimGate(**parameters.asdict())
+
+    class _PhasedFSimConverter(PointOptimizer):
```

Comment:
I don't think this should be inside the class definition

Reply:
Moved to top level. Not that there is a tight coupling between this class and the simulator; that's why I made in an inner class.